Enhance iai callgrind adapter

gamma0987 opened the pull request:
I started working on enhancing the iai-callgrind adapter.

Summary of current changes:
* Added all currently possible callgrind metrics
* `L2 Hits` will be renamed to `LL Hits` in the next iai-callgrind release
* Usage of an `Unknown` metric for metrics which are not implemented instead of failing the parser. Unknown metrics are discarded.
* The callgrind tool does not have to be enabled (for example only dhat is fine, too)
* tools can be in any order (i.e. first dhat, then callgrind)
* metrics can be in any order (first `L1 Hits` then `Instructions`)
* there is a default set of metrics for every tool that is shown, but that can be changed to anything (for example show only `Instructions`)
* There are `f64` metrics (`LLMissRate`, ...), so all metrics are parsed as `f64` instead of `u64`
* Fixed the possibility for the special values `+++inf+++` and `---inf---` in the diff percentage and/or factor
* I don't know if it's much faster but I changed some usages of `tags` with single character to `char`.

TODOs:
* Cachegrind metrics
* Error metrics for tools like memcheck, helgrind and drd

Closes #613

epompeii commented:
@gamma0987 thank you for your work on `iai-callgrind` and for your work here!

I'm going to try to find the time to dig into this code as soon as I can. When is the next release of `iai-callgrind` slated to come out?

gamma0987 commented:
I can't tell for sure, but it'll be a smaller release soon in the next days. The changes here already include the new metrics (miss and hit rates as `f64` metrics) from the main branch and the next release.

epompeii commented on lib/bencher_adapter/src/results/adapter_results.rs:
```diff
@@ -65,6 +96,35 @@ pub enum IaiCallgrindMeasure {
+    /*
+     * Unknown
+     */
+    Unknown(JsonNewMetric),
```
I really like the catch all here! Nice!

Can we make this a unit variant instead?

```suggestion
    Unknown,
```

From what I can tell, we are never using the tuple value. This also gets rid of the `Unknown` measure type definition. Maybe I'm missing something though?

epompeii commented on lib/bencher_json/src/project/measure/built_in.rs:
```diff
@@ -115,4 +149,62 @@ pub mod iai_callgrind {
     create_measure!(WritesBytes, "Writes bytes", "writes-bytes", BYTES);
+
+    // Memcheck
+    create_measure!(MemcheckErrors, "Errors", "memcheck-errors", "errors");
```
Yes, we should have different display names for the different `Errors` types.

I like your idea of separating the parse value from the final display name.

Preferably, you would create a new variant for the `create_measure!` macro that takes an additional argument, the display name.
So in the example here that would be `Memcheck Errors` or something like that.

The other macro variant would just call this longer form variant either with its parse value as the display name or with `None` (and have it be optional as you suggested).

epompeii commented:
Thank you for your work so far @gamma0987 !
Just a little bit of cleanup.

epompeii reviewed (changes requested):
Thank you for your work so far @gamma0987 !
Just a little bit of cleanup.

gamma0987 commented on lib/bencher_json/src/project/measure/built_in.rs:
```diff
@@ -99,9 +102,40 @@ pub mod iai_callgrind {
     );
+    create_measure!(GlobalBusEvents, "Ge", "global-bus-events", "events");
+
+    create_measure!(Dr, "Dr", "dr", "reads");
```
There are currently no long form names. I thought about it, but in iai-callgrind I stick to the "official" valgrind names for the momemt to avoid confusion for people who are used to the valgrind names. Also, in the output files of callgrind, ... and tools like `callgrind_annotate`, `cg_annotate`, ... the names are exactly like that. But, we could use the display string approach and give them nicer names in bencher like `Dr (Cache data reads)`, ... Is it right that the only restriction for the `JsonNewMetric::name` is a length of 64 bytes?

gamma0987 replied:
kcachegrind has long form names. So, if I am using long form names in iai-callgrind in the future I most likely will use ones very similar to kcachegrind. I think it would be nice to have long form names with a notion of the short name in bencher.

gamma0987 replied:
I'm short in time atm, so I have to postpone giving long names to another pr. It would also give me more time to think about the exact names so that they fit better with the long names I will assign in iai-callgrind in a future release.

gamma0987 commented:
I added a small fix with test in 8f8fb24 to match the new summary line which may also list regressions. In 5de87eb I added the new metrics to `QueryMeasure::get_create_or_inner`.

epompeii reviewed (approved):
@gamma0987 if you want to expand those display names, feel free to do so. Otherwise, just resolve the issue, and I'll merge this in.

gamma0987 commented:
I closed the comment for now. Thanks for your time to review this pr and for bencher.
